Guide to Spring Cloud Kubernetes (#2916)
section-engineering-education/engineering-education: content/articles/spring-cloud-k8/index.md

> ### Tutorial overview
> In this tutorial, we will learn how to use Kubernetes and Spring Cloud to build a micro-service application and integrate it with Spring Boot.

Reviewer: ```suggestion
---
layout: engineering-education
status: publish
published: true
url: /spring-cloud-k8/
title: Guide to Spring Cloud Kubernetes (K8s)
description: This tutorial will walk you through the process of creating a Spring Cloud Kubernetes application.  Spring Cloud is a Spring module that offers RAD (Rapid Application Development) functionality to the Spring framework.
author: odiwuor-amos
date: 2021-07-23T00:00:00-10:00
topics: []
excerpt_separator: <!--more-->
images:

  - url: /engineering-education/spring-cloud-k8/hero.png
    alt: Spring Cloud K8 image
---

In this tutorial, we will learn how to use Kubernetes and Spring Cloud to build a micro-service application and integrate it with Spring Boot. Spring Cloud is a Spring module that offers RAD (Rapid Application Development) functionality to the Spring framework.
<!--more-->
With the assistance of the Spring Cloud Framework, we can easily create cloud-based allocation.
```
Added Front matter.

### these commands build the docker images on minikube
cd travel-agency-service
docker build -t travel-agency-service .
cd ../client-service
docker build -t client-service .
cd ..

# sets up the secret and mongodb services
kubectl delete -f travel-agency-service/secret.yaml
kubectl delete -f travel-agency-service/mongo-deployment.yaml

kubectl create -f travel-agency-service/secret.yaml
kubectl create -f travel-agency-service/mongo-deployment.yaml

# setting up the travel-agency-service
kubectl delete -f travel-agency-service/travel-agency-deployment.yaml
kubectl create -f travel-agency-service/travel-agency-deployment.yaml

# setting up the client-service
kubectl delete configmap client-service
kubectl delete -f client-service/client-service-deployment.yaml

kubectl create -f client-service/client-config.yaml
kubectl create -f client-service/client-service-deployment.yaml

# run checks on pods to check if they are indeed running
kubectl get pods

```

### Conclusion
In this tutorial, we've seen how we can set up a Spring Cloud project and use Spring Boot and Kubernetes to integrate a micro-service platform application with each providing great performance.

> The full source code can be found [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml).

Reviewer: ```suggestion
The full source code can be found [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml).

Happy coding!

---
Peer Review Contributions by: [Eric gacoki](/engineering-education/authors/eric-gacoki/)
```
Added peer reviewer line.

Now let's add the `@EnableDiscoveryClient` and inject the `ClientController` in our controller using the `@Autowired` as shown below:

```java
@SpringBootApplication
//here is our service discovery
@EnableDiscoveryClient
public class Application {
    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }
}
//we then se it in our client controller class
@RestController
public class ClientController {
    @Autowired
    private DiscoveryClient discoveryClient;
}
```

### Creating MongoDB service
In this section, we define our MongoDB service and a configuration file for deployment. But first, we need to define both the Secret username and password to use for deployments.

```yml
# application version 1
apiVersion: v1.0.0
# we're creating a secret for our database
kind: Secret
metadata:
  name: db-secret
data:
# specifiy the password and username
  username: test==
  password: test=

```

Upon defining the configuration above, ensure you run the following command to add it to the Kubernetes cluster.

```bash
kubectl apply -f secret.yaml
```

Output:  

```bash
secret/db-secret created
```

> 
> Now proceed and add the following configurations in our deployment `YAML` file.  
> 

Reviewer: ```suggestion

Proceed to add the following configurations in our deployment `YAML` file.   

```

```yml
# api version definition
apiVersion: extensions/v1beta1
# we're deploying this application
kind: Deployment
metadata:
# creating mongodb service
  name: mongo
spec:
# with a single replica
  replicas: 1
  template:
    metadata:
      labels:
        service: mongo
      name: mongodb-service
    spec:
      containers:
      - args:
        - mongod
        - --smallfiles
        image: mongo:latest
        name: mongo
        env:
          - name: MONGO_INITDB_ROOT_USERNAME
            valueFrom:
            # these are secret values(username) we defined in the previous section
              secretKeyRef:
                name: db-secret
                key: username
          - name: MONGO_INITDB_ROOT_PASSWORD
            valueFrom:
              secretKeyRef:
              # this password is retrieved from the previous section
                name: db-secret
                key: password
```

> This deployment configuration file creates a `mongo: latest image. It then sets both of its username and password we created previously with a default`admin database.

Reviewer: ```suggestion
This deployment configuration file creates a `mongo: latest image`. It also sets the username and password that we created earlier with a default admin database.
```

### Setting up MongoDB on the Agency Service
Let's start by updating the properties to add the database credentials as shown below:

```properties
# we're updating this properties file with the db-secret that created previously
spring.cloud.kubernetes.reload.enabled=true
# setting the secret name
spring.cloud.kubernetes.secrets.name=db-secret
# the db host
spring.data.mongodb.host=mongodb-service
# specifying the port to use
spring.data.mongodb.port=27017
# The database name here is defaulted to admin
spring.data.mongodb.database=admin
# for security, we hide the username and password as they are sensitive
spring.data.mongodb.username=${MONGO_USERNAME}
spring.data.mongodb.password=${MONGO_PASSWORD}
```

> Now that we have got a complete setup of how this application works, clone the complete code on <https://replit.com/@odiwuoramos/spring-cloud> and test the application on your local machine or as well play with the code on the hosting platform.

Reviewer: ```suggestion
Now that we have understood how this application works, clone the [complete code](https://replit.com/@odiwuoramos/spring-cloud) and test it locally on your machine or play around with the code on the hosting platform.
```

### Deployment

> To deploy our application, we setup the bash script as shown below with comments for each command:  
> 

Reviewer: ```suggestion
To deploy our application, we setup a `bash script` as explained below:  

```

```bash

### first step involves building the repository
mvn clean install

### Set the docker environment for the application
eval $(minikube docker-env)

### these commands build the docker images on minikube
cd travel-agency-service
docker build -t travel-agency-service .
cd ../client-service
docker build -t client-service .
cd ..

### sets up the secret and mongodb services
kubectl delete -f travel-agency-service/secret.yaml
kubectl delete -f travel-agency-service/mongo-deployment.yaml

kubectl create -f travel-agency-service/secret.yaml
kubectl create -f travel-agency-service/mongo-deployment.yaml

### setting up the travel-agency-service
kubectl delete -f travel-agency-service/travel-agency-deployment.yaml
kubectl create -f travel-agency-service/travel-agency-deployment.yaml

### setting up the client-service
kubectl delete configmap client-service
kubectl delete -f client-service/client-service-deployment.yaml

kubectl create -f client-service/client-config.yaml
kubectl create -f client-service/client-service-deployment.yaml

# run checks on pods to check if they are indeed running
kubectl get pods

```

### Conclusion

> In this tutorial, we've seen how we can set up a Spring Cloud Kubernetes project.
> We've seen how we can use both tools to set up a microservice platform application with each providing its best `power`.
> 
> The full source code is available [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml)

Reviewer: ```suggestion
In this tutorial, we've seen how we can set up a Spring Cloud project and use Spring Boot and Kubernetes to integrate a micro-service platform application with each providing great performance.

The full source code can be found [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml).
```

> ### Tutorial overview
> 
> In this tutorial, I'll show you how to use Kubernetes and Spring cloud to build your microservice application. It'll be an interesting project that will walk you through the whole process of integrating Kubernetes with Spring Boot applications.

Reviewer: ```suggestion
### Tutorial overview
In this tutorial, we will learn how to use Kubernetes and Spring cloud to build a micro-service application and how to integrate it with Spring Boot.
```
Please remove the blank line(s) between a heading and the paragraph that follows it. Do this all through the article.

Author: @Ericgacoki ,fixed.
